scipy/special/_stirling2.py

```diff
@@ -0,0 +1,38 @@
+import numpy as np
+
+from scipy.special._ufuncs import _stirling2
+from scipy.special._ufuncs import _stirling2_inexact
+
+
+def stirling2(n, k, exact=True):
+    if exact:
+        if not (np.issubdtype(type(n), np.integer)
+                and np.issubdtype(type(k), np.integer)):
+            raise TypeError("Placeholder, need appropriate error message")
+        try:
+            output = _stirling2(n, k)
+        except TypeError:
+            output = -1
+        if output < 0:
+            return _stirling2_pyint(n, k)
```

Comment:
This begs the question: why have the `c` implementation in the first place? (e.g. does it really save that much time that the additional complexity is worth the effort?) 

A waste here is that the `_stirling2()` function does a bunch of computation that is ignored 
and then *redone* in the `_stirling2_pyint()` call. 

Couldn't we reuse some of that preliminary calculation?

Reply:
Yeah. For `exact=True` I don't think we need the C function. We really just need one for `exact=False` where `stirling2` will be a ufunc. I think we can just expose inexact from cephes with signature double f(long, long) as a private ufunc. When it has the temme approximation, it might be useful as public. 

We could reuse the work from C if the C function returns a Python list of the last array that didn't overflow, along its index. We'd want to use Cython to make working with the Python API easier. I don't think it's necessary though. As a scalar function, the pure Python version is fast enough.

Reply:
Some data: I did a very simple benchmark.

```python

from scipy.special._stirling2 import stirling2, _stirling2_pyint


def check_c():
    for n in range(26):
        for k in range(n + 1):
            stirling2(n, k)


def check_py():
    for n in range(26):
        for k in range(n + 1):
            _stirling2_pyint(n, k)


def check_inexact():
    for n in range(26):
        stirling2(n, np.arange(n + 1), exact=False)
```

```
In [2]: %timeit check_c()
764 µs ± 799 ns per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

In [3]: %timeit check_py()
2.52 ms ± 14.9 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [4]: %timeit check_inexact()
134 µs ± 181 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
```

Yeah, I don't think it's worth the added complexity to call C for `exact=True`, but the ufunc for exact=False does give a nice speedup.